pkl-core/src/main/java/org/pkl/core/resource/ResourceReaders.java

```diff
@@ -594,4 +614,114 @@ private static class FromServiceProviders {
       INSTANCE = Collections.unmodifiableList(readers);
     }
   }
+
+  private static final class External implements ResourceReader {
+    private final String scheme;
+    private final ExternalReaderProcess process;
+    private final long evaluatorId;
+    private MessageTransportResource underlying;
+
+    public External(String scheme, ExternalReaderProcess process, long evaluatorId) {
+      this.scheme = scheme;
+      this.process = process;
+      this.evaluatorId = evaluatorId;
+    }
+
+    private MessageTransportResource getUnderlyingReader()
+        throws ExternalReaderProcessException, IOException {
+      if (underlying != null) {
+        return underlying;
+      }
+
+      var spec = process.getResourceReaderSpec(scheme);
+      if (spec == null) {
+        throw new ExternalReaderProcessException(
+            ErrorMessages.create("externalReaderDoesNotSupportScheme", "resource", scheme));
+      }
+      underlying =
+          new MessageTransportResource(
+              spec, new ExternalResourceResolver(process.getTransport(), evaluatorId));
+      return underlying;
+    }
+
+    @Override
+    public String getUriScheme() {
+      return scheme;
+    }
+
+    @Override
+    public boolean hasHierarchicalUris() throws ExternalReaderProcessException, IOException {
+      return getUnderlyingReader().hasHierarchicalUris();
+    }
+
+    @Override
+    public boolean isGlobbable() throws ExternalReaderProcessException, IOException {
+      return getUnderlyingReader().isGlobbable();
+    }
+
+    @Override
+    public Optional<Object> read(URI uri) throws IOException, ExternalReaderProcessException {
+      return getUnderlyingReader().read(uri);
+    }
+
+    @Override
+    public boolean hasElement(SecurityManager securityManager, URI elementUri)
+        throws IOException, SecurityManagerException, ExternalReaderProcessException {
+      return getUnderlyingReader().hasElement(securityManager, elementUri);
+    }
+
+    @Override
+    public List<PathElement> listElements(SecurityManager securityManager, URI baseUri)
+        throws IOException, SecurityManagerException, ExternalReaderProcessException {
+      return getUnderlyingReader().listElements(securityManager, baseUri);
+    }
+
+    @Override
+    public void close() {
+      process.close();
+    }
+  }
+
+  private static final class MessageTransportResource implements ResourceReader {
```

Comment:
For consistency; no other readers have `Reader` name postfixes.

```suggestion
  private static final class MessageTransport implements ResourceReader {
```